Refactor build system, simplify templates and escape macros in code

johnd0e opened the pull request:
- Refactor build system into several modules (can be used as cli utilities). Implement `watch` mode to auto-rebuild on source changes.
- Escape macros in code in order to keep js-validity, fix #50.
  Use template from settings instead of url harcoding (closes #150).
- Simplify userscripts source template, rename plugin sources:  `*.user.js` -> `*.js`, rearrange source directories.
- Remove timestamp component from `version` of Release scripts (closes #99).
- Additional dev tools: `web_meta_gen.py`, `web_server_local.py`, `tampermonkey_stubs_gen.py`.

johnd0e commented:
.

modos189 commented on build.py:
```diff
@@ -20,19 +20,18 @@
+defaultBuild = None
 # load option local settings file
-try:
+if os.path.isfile('./localbuildsettings.py'):
```
If file is missing, there will be an exception. Not great, not terrible.

BTW, we can use:
```
# load option local settings file and default build
try:
    from localbuildsettings import buildSettings as localBuildSettings, defaultBuild
    buildSettings.update(localBuildSettings)
except ImportError:
    localBuildSettings = {}
    defaultBuild = None
```

johnd0e commented:
I am also going to enhance build script to be able to build single plugin (even 3rd-party).

Commit:
Refactor build system, simplify templates and escape macros in code

Refactor build system into several modules:
- `settings.py`: to provide common settings (from `buildsettings.py` / `localbuildsettings.py`)
- `build_plugin.py`: to build single plugin
  formats userscript metablock, preprocesses sources - makes templates substitutions,
  inlines dependencies js/css/img - and wraps into wrapper (from `pluginwrapper.py`)
  - `pluginwrapper_noinject.py` - alternative wrapper, useful for debugging scripts.
     Sample usage in `buildsettings.py`, for build name 'tmdev'.
- `build_mobile.py`: to build android apk (embedding scripts from given directory)
- `build.py`: to build all: main IITC script, all plugins, and (optionally) Android apk.
  - `--watch` mode (auto-rebuild on sources changes)

Each module can be used independently as cli utility.

Escape macros in code in order to keep js-validity, fix #50.
Simplify userscripts source template, completely get rid of these ugly substitutions:
- `@@METAINFO@@`
- `@@PLUGINSTART@@`
- `@@PLUGINEND@@`
- `.@@DATETIMEVERSION@@`
- `[@@BUILDNAME@@-@@BUILDDATE@@]`

Close #150: use template from settings instead of url harcoding.

Close #99: remove timestamp component from `version` of Release scripts.

modos189 commented on code/boot.js:
```diff
@@ -829,23 +829,23 @@ window.registerMarkerForOMS = function(marker) {
 
 try {
-@@INCLUDERAW:external/autolink-min.js@@
+@include_raw:external/autolink-min.js@
```
Maybe we should add quotes, make a string?

modos189 commented on plugins/debug-raw-portal-data.user.js:
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-raw-portal-data
```
Did you leave your id here on purpose, or did you forget to remove it?

modos189 commented on plugins/periodic-refresh.user.js:
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-periodic-refresh
```
This is probably where you forgot to remove ID

modos189 commented on total-conversion-build.js:
```diff
@@ -1,13 +1,10 @@
+// @description    Total conversion for the ingress intel map.
 // @run-at         document-end
-// ==/UserScript==
 
```
BTW, maybe we should move this file to `code` folder?

modos189 commented on plugins/fix-china-map-offset.user.js:
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-fix-china-map-offset
+﻿// @id             iitc-plugin-fix-china-map-offset
```
again

modos189 commented on build_plugin.py:
```diff
@@ -0,0 +1,182 @@
+
+    script = re.sub(r"'@bundle_code@';", bundle_code, script)
+    wrapper = import_module(settings.plugin_wrapper)
+    template = r"'@(\w+)(?::([0-9a-zA-Z_./-]+))?@'"
```
Ooh. Uh...)
At least add a comment, please, that it's for finding `@type:string@` tags

modos189 commented on build_plugin.py:
```diff
@@ -0,0 +1,182 @@
+        append_line('match', settings.match)
+    append_line('grant', 'none')
+    meta.append('// ==/UserScript==\n')
+    return '\n'.join(meta), is_plugin
```
Please add blank lines, everything is now merging into a solid block

modos189 commented on plugins/basemap-gaode.user.js:
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-basemap-gaode
+﻿// @id             iitc-plugin-basemap-gaode
```
id?

modos189 commented on plugins/basemap-kartverket.user.js:
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-basemap-kartverket
+﻿// @id             iitc-plugin-basemap-kartverket
```
id?

modos189 commented on plugins/link-show-direction.user.js:
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-link-show-direction
```
id?

modos189 reviewed (dismissed):
It's pretty cool on the whole. I left a few questions in the comments

Commit:
plugins/deleted: remove old stubs

Commit:
web_meta_gen.py: utility to generate meta.json (for IITC-Button)

Commit:
web_server_local.py: utility to start local webserver for specified build name

Commit:
clear excessive permissions

Commit:
smartphone.css: move from mobile/ to root

Commit:
build scripts: drop Python 2.7 support

From now Python at least version 3.6 required

Commit:
speech-search -> deleted/

It's unfunctional since underlying webkit API removal (several years ago)

Commit:
plugins: *.user.js -> *.js

*.js for raw sources
*.user.js for ready-to-use scripts

Commit:
update .travis.yml

Commit:
sources rearrangement

external -> /core/external/,        /plugins/external/
images   -> /core/images/,          /plugins/images/
         -> /core/external/images/, /plugins/external/images/

Commit:
build.py: improve watch mode, and add same possibility to build_plugin.py

Watch mode is now properly implemented:
it considers exact source dependencies
(instead of reacting on just any directory change)

Commit:
settings.py: tie source/target directories to script's directory (instead of cwd)

Can be overriden in localbuildsettings.py

Commit:
web_meta_gen.py: support (multilingual) descriptions for categories

Descriptions source: plugins/categories.json
(from site's repo: https://github.com/IITC-CE/iitc.modos189.ru/blob/master/box.py)

Generated meta.json now used for:
- site generation
- IITC-Button

modos189 commented on plugins/bookmarks.css:
```diff
@@ -280,7 +280,7 @@
 }
 #bkmrksTrigger, .bkmrksStar span, .portal-list-bookmark span {
-	background-image:url(@@INCLUDEIMAGE:images/bookmarks-img.png@@);
+	background-image:url(images/bookmarks-img.png);
```
Should we use `@include_img` here?

johnd0e replied:
No, that is handled in another way.
Urls in CSS are easy to extract with regexp, so no need for special templates.  

See corresponding change in `bookmarks.js`.

modos189 replied:
Ok

modos189 reviewed (changes requested):
Good job!

Commit:
build.py: minor correction of watch mode

Do not try to track changes of localbuildsettings.
Anyway it had no effect without settings reloading.

Commit:
tampermonkey_stubs_gen.py: new helper utility

to make Tampermonkey read userscripts from local disk.
